indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -1271,6 +1277,102 @@ public void remove(final Task task)
     }
   }
 
+  /**
+   * Acquire a read lock to perform the segment transactional append action for a given datasource.
+   * Also verifies that all the locks are of the type APPEND for the task.
+   * @param task task to perform the append action
+   * @param lockAcquireTimeoutMillis milliseconds to wait for lock acquisition
+   */
+  public void acquireTransactionalAppendLock(Task task, long lockAcquireTimeoutMillis)
+  {
+    final String datasource = task.getDataSource();
+    final ReentrantReadWriteLock readWriteLock =
+        datasourceToConcurrentLock.computeIfAbsent(datasource, ds -> new ReentrantReadWriteLock(true));
+
+    synchronized (readWriteLock) {
+      try {
+        final boolean acquired = readWriteLock.readLock()
+                                              .tryLock(lockAcquireTimeoutMillis, TimeUnit.MILLISECONDS);
+        if (!acquired) {
+          throw DruidException
+              .forPersona(DruidException.Persona.OPERATOR)
+              .ofCategory(DruidException.Category.TIMEOUT)
+              .build("Timed out while acquiring transactional append lock for datasource[%s].", datasource);
+        }
```

Comment:
This can be outside the try and synchronized.

Reply:
Done, thanks